lib/elixir/src/elixir_erl.erl

```diff
@@ -559,6 +581,38 @@ get_type_docs(Set, Types) ->
      {_, Line, Doc} <- ets:lookup(Set, {Kind, Tuple})]
   ).
 
+translate_doc({{Name, Arity}, Line, Kind, Signature, Doc}) ->
+  {{translate_doc_kind(Kind), Name, Arity},
+   erl_anno:new(Line),
+   [signature_to_binary(Name, Signature)],
+   doc_value(Doc),
+   #{signature => Signature}
+  };
+translate_doc({{Name, Arity}, Line, Kind, Doc}) ->
+  {{translate_doc_kind(Kind), Name, Arity},
+   erl_anno:new(Line),
+   [],
+   doc_value(Doc),
+   #{}
+  }.
+
+translate_doc_kind(def)           -> function;
+translate_doc_kind(callback)      -> callback;
+translate_doc_kind(type)          -> type;
+translate_doc_kind(opaque)        -> opaque;
+translate_doc_kind(defmacro)      -> macro;
+translate_doc_kind(macrocallback) -> macrocallback.
```

Comment:
@josevalim are you happy with this mapping? I was unsure about whether to use `type` for `opaque` too, but decided to use `opaque` as that's what we're using in `ExDc` at the moment.

Reply:
I also thought about adding a generic clause that returns the argument, but in this case I think it's better not to extend the kinds silently if a new one is introduced. Instead a function clause error would provide an opportunity to make a conscious decision about naming.

Reply:
Wouldn't it be clearer to have:

```erlang
translate_doc_kind(def)           -> function;
translate_doc_kind(defmacro)      -> macro;
translate_doc_kind(Type) -> Type.
```

I know that this isn't exactly the same, but it would allow seamless expansion of supported kinds in future without changes there.

Reply:
@hauleth I think it's better to let new kinds fail on this first so that there is a conscious decision about how to name them in the docs chunk.

Reply:
Yeah, let's go the explicit road for now. :+1: